### ISSUE TITLE
Emoji in header breaks generated link

### ISSUE THREAD

**Issue by adrianmcli** [thlorenz/anchor-markdown-header#36]
Continued from https://github.com/thlorenz/doctoc/issues/123

In short: **If you place an emoji into a header, the generated anchor tag does not work.**

----

The actual link generated by GitHub just leaves out the emoji, but has a dash in there to show the space.

So I actually tried it out with a heading like this:

```
# Modules 📦
```

And the Markdown required is:

```
- [Modules](#modules-)
```

You can see it here: https://github.com/adrianmcli/next-boilerplate/blob/master/README.md

**Comment by adrianmcli** [comment]
Of course, you could also put the emoji inside the anchor link text as well, but I thought it didn't look good for my use case.

For the sake of posterity, it would look like this:

```
- [Modules 📦](#modules-)
```

**Comment by thlorenz** [comment]
OK thanks for that info, so what we'd need to do is replace any emoj with a dash as well.
Could you research how to detect any emoj in a regex? I've got no clue ;)

You can test things by forking and modifying this repo.
You'll get the quickes feedback if you just add a failing test here with an emoj.
Then you just try to modify the used regex until the test passes.

Finally you can PR with your changes. I'd greatly appreciate it.
Thanks.

**Comment by adrianmcli** [comment]
I don't think it becomes a space. It kind of just disappears. The dash is there because I had a space in between the emoji and the word.

For example, this:

```
# Modu📦les
```

Would become

```
- [Modu📦les](#modules)
```

**Comment by adrianmcli** [comment]
So we need to:

1. Include the emoji as if it was an actual character/word.
2. Convert spaces to dashes (regular conversion).
3. Strip out the emoji.

**Comment by thlorenz** [comment]
Sounds good .. basically just add tests that assume it's doing all that and then make'em pass.

Quite easy really :P main challenge how do you regex match emojis

**Comment by adrianmcli** [comment]
Apparently, detecting emojis with regex is actually really hard. Maybe this package can help: https://github.com/mathiasbynens/emoji-regex

**Comment by adrianmcli** [comment]
@thlorenz PR submitted, yay! #37

**Comment by anoff** [comment]
There seem to be some emojis that are not completely captured by this solution
![image](https://user-images.githubusercontent.com/7142618/29278745-f0aeed34-8115-11e7-9c2b-b404e405cc3d.png)

Didn't get around to analyze what makes this one special. Possibly related to the issue in the library you're using https://github.com/mathiasbynens/emoji-regex/issues/28
`⚒`

**Comment by thlorenz** [comment]
Hope we can catch those as well .. don't have bandwidth to attack this myself, but maybe you can help (PRs appreciated).
Also @mathiasbynens should be considered the emoji regexpert (new word) so whatever he says in his issue would very likely be correct :)

**Comment by mathiasbynens** [comment]
> Didn't get around to analyze what makes this one special. Possibly related to the issue in the library you're using mathiasbynens/emoji-regex#28

Yeah, that’s likely the issue. emoji-regex follows the Unicode standard, detecting only official emoji sequences. Apple’s macOS emoji picker randomly inserts U+FE0F after certain emoji despite that resulting in a non-standard sequence.

Why would you want to strip emojis, though? They’re perfectly valid in IDs and `#foo`-style in-page anchors. IMHO, a better fix would leave emojis intact and make sure the links are working instead.

**Comment by adrianmcli** [comment]
@mathiasbynens unfortunately, that's just how the GitHub README links work.

```
## my title🕵️here
```

Generates this anchor:

```
#my-titlehere
```

**Comment by anoff** [comment]
tbh I can live with the current behaviour. Just wanted to add the comment here for future generations to see. Depending on whether the issue gets fixed in the underlying library I'd recommend a _Known Issues_ in the readme though :) I'll try to stick around and PR the docs if needed.

**Comment by Miserlou** [comment]
This never got resolved! Still breaks for me. Can we just drop the emoji or give me an option to?

I use emojis in headers here: https://github.com/Miserlou/dnd-tldr

**Comment by ccheever** [comment]
@Miserlou I think your links will work if you just remove the emojis from the links. I tried modifying the URL fragment on one of your broken links and it worked.

**Comment by ctsstc** [comment]
If I remove the emojis from the links then it breaks in other apps like VSCode's markdown preview. I would like it to work on Github and other apps 🤔

**Comment by basnijholt** [comment]
The biggest problem is that the behavior is inconsistent. For example
`# Alarm clock ⏰` has `#alarm-clock-` as link
`# Apple Watch ⌚️` has `#apple-watch-%EF%B8%8F` as link.

This gets quite annoying when automatically generating table of contents, such as I do here: https://github.com/basnijholt/home-assistant-config/blob/35f3ae3942c5d343efe133fccd85415d4bdf6501/README.md#automations---table-of-content

**Comment by jon424** [comment]
Is there any news on this issue? This works in VSCode for me:
`[React](#⚛️-React)`

but doesn't on github. It tries to access this URL: 
`#⚛%EF%B8%8F-React`

**Comment by Alynva** [comment]
I'm getting these `%EF%B8%8F` unicodes that prints as empty chars. In the inspector I see this:

![image](https://user-images.githubusercontent.com/12958674/135875064-4e6924c0-11fd-487c-9d72-e3372146ade4.png)

But hovering the link it shows this:

![image](https://user-images.githubusercontent.com/12958674/135875167-263e7ff5-1d83-4d83-935c-a86ce0dca9e3.png)

The same that goes to the URL:

![image](https://user-images.githubusercontent.com/12958674/135875327-12e92a5e-25d7-4a55-9c7e-ef54b4f8a295.png)

That is hanpping with multiple emojis: `⚠️✍️⏲️🛠️⚙️☝️⚡️` and others...

But as other said, I prefer that the emojis was just added as is, not removing like that because it breaks on other softwares.

**Comment by kevin-david** [comment]
#45 should address this by taking a new version of [`emoji-regex`](https://www.npmjs.com/package/emoji-regex)
